### PR TITLE
Single inheritance

### PR DESCRIPTION
**Pull request by kunjsong01**
https://github.com/esbmc/esbmc/pull/854 is too large and messy for review. 

Based on previous discussion on slack, this is a PR separated from https://github.com/esbmc/esbmc/pull/854, adding support for single inheritance only. 

Features implemented: 
1. Populated class type's `components` and `methods` vectors (which will be used for polymorphism support, e.g. generating implicit ctor/dtor code for vptr initializations).
2. Added base initializer conversion and generate a type casted argument for base initializer.
3. Added the corresponding test cases.

### CODE REVIEW

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 1]
```diff
@@ -1024,3 +1097,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+}
+
+void clang_cpp_convertert::gen_typecast_base_ctor_call(
+  const exprt &callee_decl,
+  side_effect_expr_function_callt &call,
+  exprt &initializer)
+{
+  // sanity checks
+  assert(initializer.derived_this_arg() != "");
+  assert(initializer.base_ctor_derived());
```
Comment (kunjsong01) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1206]: @mikhailramalho 
This function would appear to be adjustment only. 
Shall we move it to adjuster?

File: regression/esbmc-cpp/inheritance_bringup/single_member_fail/main.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+public:
+  int data;
+  Vehicle() { data = 1; }
+};
+
+class Motorcycle: public Vehicle
+{
+public:
+  Motorcycle() : Vehicle() {}
+  //Motorcycle() {}
```
Comment (lucasccordeiro) [inline regression/esbmc-cpp/inheritance_bringup/single_member_fail/main.cpp]: Remove this line.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 3]
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
 
   return nullptr;
 }
+
+bool clang_c_convertert::check_alignment_attributes(
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:3278]: You should comment on this method.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 4]
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
...
+  const clang::FieldDecl *field,
+  struct_typet::componentt &comp)
+{
+  if(field->hasAttrs())
+  {
+    const auto &attrs = field->getAttrs();
+    for(const auto &attr : attrs)
+    {
+      if(attr->getKind() == clang::attr::Aligned)
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_convert.cpp:3287]: Add explanation.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 5]
```diff
@@ -220,7 +221,19 @@ bool clang_cpp_convertert::get_function(
   if(fd.isDependentContext())
     return false;
 
-  return clang_c_convertert::get_function(fd, new_expr);
+  // Only convert non-trivial functions, e.g. non-trivial ctor/dtor, copy ctor, move ctor .etc
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Remove this block of comments.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 6]
```diff
@@ -220,7 +221,19 @@ bool clang_cpp_convertert::get_function(
...
+  //if(fd.isTrivial())
+  //  return false;
+
+  if(clang_c_convertert::get_function(fd, new_expr))
+    return true;
+
+  // add additional annotations for class/struct/union methods
+  if(const auto *md = llvm::dyn_cast<clang::CXXMethodDecl>(&fd))
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:228]: Shall we write it as an AND statement?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 7]
```diff
@@ -268,23 +282,60 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
...
+      cxxrd->getNameAsString(),
+      cxxrd->getNumBases(),
+      cxxrd->getNumVBases());
+
+    // TODO-split: add methods from base class
+    for(auto base : cxxrd->bases())
+    {
+      // `base` type is clang::CXXBaseSpecifier
+      base.getType().dump();
+      //log_error("inheritance is not supported in {}", __func__);
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Shall we remove these lines?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 8]
```diff
@@ -321,6 +372,10 @@ bool clang_cpp_convertert::get_struct_union_class_methods(
     {
       // Add only if it isn't static
       if(!cxxmd->isStatic())
+      {
```
Comment (lucasccordeiro) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: Remove brackets?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
Some minor comments.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 6 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:228]: It checks to see if the operand is of the specified type, and if so, returns a pointer to it, otherwise returns a nullptr.  

This is part of the `clang_cpp_convertert::get_function` which is used to convert both ordinary functions and class methods. So we might get an ordinary function rather than the non-CXXMethodDecl for class method, in which case we'll get a nullptr.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 7 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp]: Removed, this will be part of the next PR "split-polymorphism".

**Review by lucasccordeiro — APPROVED** [review]
LGTM.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 9]
```diff
@@ -298,13 +303,22 @@ bool clang_c_convertert::get_struct_union_class(const clang::RecordDecl &rd)
...
+  //       A C++ union can have member functions but not virtual functions.
+  //       Just use struct_typet for C++?
   struct_union_typet t;
   if(rd.isUnion())
     t = union_typet();
   else
     t = struct_typet();
   t.tag(name);
 
+  if(rd.isClass())
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp]: is this needed?

File: regression/esbmc-cpp/inheritance_bringup/single_member/main.cpp [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+public:
+  int data;
+  Vehicle() { data = 1; }
+};
+
+class Motorcycle: public Vehicle
+{
+public:
+  Motorcycle() : Vehicle() {}
+  //Motorcycle() {}
```
Comment (mikhailramalho) [inline regression/esbmc-cpp/inheritance_bringup/single_member/main.cpp]: remove

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 11]
```diff
@@ -433,7 +418,7 @@ bool clang_c_convertert::get_struct_union_class_fields(
 
 bool clang_c_convertert::get_struct_union_class_methods(
   const clang::RecordDecl &,
-  struct_union_typet &)
+  struct_typet &)
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp:415]: why this change? I think unions can have methods.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 12]
```diff
@@ -3038,7 +3027,7 @@ std::string clang_c_convertert::get_decl_name(const clang::NamedDecl &nd)
   return rso.str();
 }
 
-static std::string
+std::string
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp:3024]: this can be an inline static function, why you removed it?

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 13]
```diff
@@ -3291,3 +3280,51 @@ clang_c_convertert::get_top_FunctionDecl_from_Stmt(const clang::Stmt &stmt)
...
+          log_error("{}", oss.str());
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+bool clang_c_convertert::is_field_global_storage(const clang::FieldDecl *field)
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.cpp:3318]: do we really need a separate function for this?

File: src/clang-c-frontend/clang_c_convert.h [hunk 14]
```diff
@@ -38,8 +38,12 @@ class StringLiteral;
 class IntegerLiteral;
 class FloatingLiteral;
 class TagDecl;
+class FieldDecl;
 } // namespace clang
 
+extern std::string
+getFullyQualifiedName(const clang::QualType &, const clang::ASTContext &);
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_convert.h:45]: This doesn't need to be extern, since its definition is in the cpp file. 

Also, why do you need this?

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 13 above)
Comment (mikhailramalho) [reply src/clang-c-frontend/clang_c_convert.cpp:3318]: I see why you need it now: to reuse in the cpp convert. Please use it in clang_cpp_convertert::get_struct_union_class_fields, in the loop after `  // let's check for (virtual) base classes if this declaration is a CXXRecordDecl,`

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 15]
```diff
@@ -268,23 +278,58 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
...
+      "Class {} has {} bases, {} virtual bases",
+      cxxrd->getNameAsString(),
+      cxxrd->getNumBases(),
+      cxxrd->getNumVBases());
+
+    // TODO-split: add methods from base class
+    for(auto base : cxxrd->bases())
+    {
+      // `base` type is clang::CXXBaseSpecifier
+      base.getType().dump();
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: cleanup this.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 16]
```diff
@@ -268,23 +278,58 @@ bool clang_cpp_convertert::get_struct_union_class_fields(
...
+
+    // Check for alignment attributes
+    if(check_alignment_attributes(field, comp))
+      return true;
+
+    // Don't add fields that have global storage (e.g., static)
+    if(is_field_global_storage(field))
+      continue;
+
+    // [C++ annotation]: set parent in component's type
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:296]: do we need to set this field?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 17]
```diff
@@ -807,8 +866,14 @@ bool clang_cpp_convertert::get_function_body(
         }
         else
         {
-          log_error("Base class initializer is not supported in {}", __func__);
-          abort();
+          // TODO-split: add base class initializer
+          init->getInit()->dump();
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp]: cleanup this

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 18]
```diff
@@ -807,8 +866,14 @@ bool clang_cpp_convertert::get_function_body(
...
+          // Add additional annotation for `this` parameter
+          initializer.derived_this_arg(
+            ftype.arguments().at(0).get("#identifier"));
+          initializer.base_ctor_derived(true);
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:873]: why do we need to call derived_this_arg and base_ctor_derived?

I think the get_expr in the next line will overwrite anything set in this initializer object.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 19]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+  std::string access;
+  if(get_access_from_decl(field, access))
+    return true;
+
+  // annotate access in component
+  comp.set_access(access);
+  return false;
+}
+
+bool clang_cpp_convertert::get_access_from_decl(
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1105]: do we actually need this? We are not doing any typecheck in the IR, so the access information is useless for us (clang already check if not code call violates access definitions).

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 20]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+  {
+    typet rtn_type("constructor");
+    component_type.return_type() = rtn_type;
+  }
+
+  // annotate parent
+  std::string parent_class_name = getFullyQualifiedName(
+    ASTContext->getTagDeclType(cxxmdd->getParent()), *ASTContext);
+  std::string parent_class_id = tag_prefix + parent_class_name;
+  component_type.set("#member_name", parent_class_id);
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1158]: do we need this information?

Also, instead of exporting getFullyQualifiedName, you should call `get_decl(cxxmdd->getParent())` and access its name.

I fail to see why we need this code tho.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 21]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+
+  // annotate name
+  std::string method_id, method_name;
+  clang_c_convertert::get_decl_name(*cxxmdd, method_name, method_id);
+  new_expr.name(method_id);
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1163]: do we need this?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 22]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+
+  // annotate access
+  std::string access;
+  if(get_access_from_decl(cxxmdd, access))
+    return true;
+  new_expr.set("access", access);
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1169]: do we need this?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 23]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+
+  // annotate base name
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1171]: These two lines feel wrong...

base_name = unique identifier
pretty_name = name printed when we use --goto-functions-*, --symbol-table-*, --program-* and others

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 24]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+  new_expr.base_name(parent_class_name);
+
+  // annotate pretty name
+  new_expr.pretty_name(parent_class_name);
+
+  // annotate inline
+  new_expr.set("is_inlined", cxxmdd->isInlined());
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1178]: do we need this?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp [hunk 25]
```diff
@@ -1024,3 +1089,138 @@ bool clang_cpp_convertert::get_decl_ref(
...
+
+  // annotate location
+  locationt location_begin;
+  get_location_from_decl(*cxxmdd, location_begin);
+  new_expr.location() = location_begin;
+
+  // We still need to add a trivial ctor/dtor as a `component` in class symbol's type
+  // remove "statement: skip" otherwise it won't be added
+  // TODO-split: Any more elegant way to do it?
```
Comment (mikhailramalho) [inline src/clang-cpp-frontend/clang_cpp_convert.cpp:1187]: Can you explain why we need this code?

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 1 above)
Comment (mikhailramalho) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1206]: It's hard to tell... can you explain why we need this function?

File: src/util/irep.cpp [hunk 26]
```diff
@@ -538,6 +538,8 @@ const irep_idt irept::a_object_type = dstring("object_type");
 const irep_idt irept::a_cmt_size = dstring("#size");
 const irep_idt irept::a_cmt_type = dstring("#type");
 const irep_idt irept::a_type_id = dstring("typeid");
+const irep_idt irept::a_derived_this_arg = dstring("#derived_this_arg");
```
Comment (mikhailramalho) [inline src/util/irep.cpp:541]: why do we need this?

File: src/util/irep.h [hunk 27]
```diff
@@ -264,6 +264,16 @@ class irept
     return get(a_pretty_name);
   }
 
+  inline const irep_idt &derived_this_arg() const
```
Comment (mikhailramalho) [inline src/util/irep.h:267]: ditto

File: src/util/irep.h [hunk 28]
```diff
@@ -940,6 +950,16 @@ class irept
     set(a_pretty_name, val);
   }
 
+  inline void derived_this_arg(const irep_idt &val)
```
Comment (mikhailramalho) [inline src/util/irep.h:953]: ditto

File: src/util/irep.h [hunk 29]
```diff
@@ -1232,6 +1252,8 @@ class irept
   static const irep_idt a_end_location, a_guard, a_label, a_lhs, a_location;
   static const irep_idt a_object_type, a_cmt_size, a_cmt, a_type_id;
   static const irep_idt a_cmt_type;
+  // annotations for typecasting derived derived class `this` to base class type
```
Comment (mikhailramalho) [inline src/util/irep.h:1255]: ditto

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
@kunjsong01 I added several comments, including a lot of nitpick. Remember to remove unnecessary code (calls to .dump()) and run clang-format for the next version).

Please, double-check all the new attributes we are setting and if they are actually needed.

Your patches are getting increasingly harder to review (that's why I asked you to explain some of your decisions). This is a good sign.

File: src/clang-c-frontend/clang_c_convert.cpp [hunk 30]
```diff
@@ -298,13 +303,22 @@ bool clang_c_convertert::get_struct_union_class(const clang::RecordDecl &rd)
   if(context.find_symbol(id) != nullptr)
     return false;
 
+  // TODO: Fix me when we have a test case using C++ unioin.
```
Comment (rafaelsamenezes) [inline src/clang-c-frontend/clang_c_convert.cpp]: unioin =>  union

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]
Could you add more test-cases? For: static fields, nested inheritance, and pointer casts .

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 9 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: I suppose so... just need to distinguish class from struct, making it more intuitive in the symbol table, otherwise it would mark everything `struct`:
Screenshot 2023-01-25 at 12 51 00

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 11 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:415]: > why this change?

Need to use the `methods()` as in `type.methods().push_back(comp);`. Our `struct_union_typet` class doesn't have this function.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 9 above)
Comment (mikhailramalho) [reply src/clang-c-frontend/clang_c_convert.cpp]: If it's just to print "class" instead of "struct", we don't really need it.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 11 above)
Comment (mikhailramalho) [reply src/clang-c-frontend/clang_c_convert.cpp:415]: Do we need to add the methods to the type? i.e., is there anywhere in esbmc after the frontend (goto, ssa, smt), that we call methods() to get the list of methods?

If it's only to add the methods to the type, I'd say we don't need it.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 9 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: done removed

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 11 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:415]: Yes, to generate the implicit code for constructors for vptr initializations.

File: src/clang-c-frontend/clang_c_convert.h (hunk 14 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.h:45]: Done removed `extern`. 

>why do you need this?

Need to be used in clang-cpp-converter to get the parent_class_id

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 15 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp]: done

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 13 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp:3318]: >Please use it in clang_cpp_convertert::get_struct_union_class_fields, in the loop after   // let's check for (virtual) base classes if this declaration is a CXXRecordDecl,

Yep, it's being used in the `//Parse the fields` loop after the that base loop.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 17 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp]: Done removed

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 18 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:873]: >why do we need to call derived_this_arg and base_ctor_derived?

In short, the purpose is to generate typecasted `this` argument when calling base ctor in derived ctor, something like:
```cpp
 91 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 92
 93 Motorcycle (c:@S@Motorcycle@F@Motorcycle#):
 94         // 32
 95         FUNCTION_CALL:  Vehicle(( Vehicle *)this)  // <---- typecasted `this` argument
 96         // 33 file main.cpp line 13 function Motorcycle
 97         END_FUNCTION // Motorcycle
 98 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
"derived_this_arg" == symbol id for Motorcycle `this`
"base_ctor_derived" means "we are implicitly calling the base ctor in derivied ctor" 
(sorry about the wacky naming as I wan't able to give it a more intuitive and compact name :sweat_smile:) 

>I think the get_expr in the next line will overwrite anything set in this initializer object.

Correct... but we'll use them before the overwriting (i.e. the swapping) in `clang_cpp_convertert::get_constructor_call`. 

P.S. Combined with the new function `clang_cpp_convertert::gen_typecast_base_ctor_call`, I found these flags work pretty well when dealing with multiple inheritances (we'll see more test cases enabled for multiple inheritance in https://github.com/esbmc/esbmc/pull/884 when this PR is merged). Basically it's able to generate: 
```cpp
127 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
128 // given Rectangle class derives from Shape and PaintCost classes
129 Rectangle (c:@S@Rectangle@F@Rectangle#):
130         // 40
131         FUNCTION_CALL:  Shape(( Shape *)this)
132         // 41
133         FUNCTION_CALL:  PaintCost(( PaintCost *)this)
134         // 42 file main.cpp line 28 function Rectangle
135         END_FUNCTION // Rectangle
136 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 23 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1171]: I see... thanks very much for the explanations. I mistakes base_name for the parent class id. These annotations are added to match the symbol types generated by the old frontend. 


Now fixed: 
base_name = method id
pretty_name = method name

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 25 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1187]: Preparation for vptr and vtable extensions. We need to make some implicit code for vptr later - sorry it may not be obvious for the time being.

File: src/clang-c-frontend/clang_c_convert.cpp (hunk 30 above)
Comment (kunjsong01) [reply src/clang-c-frontend/clang_c_convert.cpp]: done fixed.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 20 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1158]: >Also, instead of exporting getFullyQualifiedName, you should call get_decl(cxxmdd->getParent()) and access its name.

well I thought of using `get_decl(cxxmdd->getParent(), new_expr)` ...  but it would appear to go through lots of recursions, seems to be an overkill and might be quite head scratching when we need to debug in the future (e.g. dealing with 25 frames in gdb). Obviously, here we just want the to get the id and annotate `member_name`. Using the getFullyQualifiedName function appears to be a neat way to do so.

File: src/clang-cpp-frontend/clang_cpp_convert.cpp (hunk 1 above)
Comment (kunjsong01) [reply src/clang-cpp-frontend/clang_cpp_convert.cpp:1206]: >can you explain why we need this function?

to generate typecasted `this` argument when calling base ctor in derived ctor. 
Please refer to the comment https://github.com/esbmc/esbmc/pull/882#discussion_r1088164013

**Comment by kunjsong01** [comment]
Hi @mikhailramalho , thanks a lot for the insightful reviews (and the explanations). I've fixed most of them. 

>do we need this? 

Good question.  
Honestly I also feel that I've added more annotations than possibly required. It would appear that adding the additional annotations might seem unnecessary, and the purpose is not quite obvious .

It's difficult to say for sure whether we need them or not. I just added these annotations to match the symbol table as much as I can. IIRC, some of them are needed to generate vtable and vptr. I may need them to generate implicit code in ctors/dtors for vptr stuff (this will become another PR after https://github.com/esbmc/esbmc/pull/884).

Apologies for not being able to give a concrete answer to this question. In order to speed up the develoment, I'm currently adopting a "learn-and-try" style development. As the development moves forward, it would become clearer. 

For the time being, I'd suggest let's leave it be and focus on pushing the development to pass more test cases (and yes we have more TCs passing in https://github.com/esbmc/esbmc/pull/884 :smiley:). But please be reassured that I'll purge these annotations after passing the virtual method test cases, and also please be assured that for any additional annotation to be added I'll try to confine them in a function so that it would be easy to indentify the redundant in the future.

**Comment by kunjsong01** [comment]
> Could you add more test-cases? For: static fields, nested inheritance, and pointer casts .

Hi @rafaelsamenezes thanks very much for the review. 
Good point. But this would be out of scope for this PR. I've got more test cases passing in the next PR https://github.com/esbmc/esbmc/pull/884 (which is currently gated by this PR) :smiley:. Currently, the focus of development is to pass the inheritance test suite in esbmc-cpp/inheritance.

**Comment by mikhailramalho** [comment]
> Hi @mikhailramalho , thanks a lot for the insightful reviews (and the explanations). I've fixed most of them.
> 
> > do we need this?
> 
> Good question. Honestly I also feel that I've added more annotations than possibly required. It would appear that adding the additional annotations might seem unnecessary, and the purpose is not quite obvious .
> 
> It's difficult to say for sure whether we need them or not. I just added these annotations to match the symbol table as much as I can. IIRC, some of them are needed to generate vtable and vptr. I may need them to generate implicit code in ctors/dtors for vptr stuff (this will become another PR after #884).
> 
> Apologies for not being able to give a concrete answer to this question. In order to speed up the develoment, I'm currently adopting a "learn-and-try" style development. As the development moves forward, it would become clearer.
> 
> For the time being, I'd suggest let's leave it be and focus on pushing the development to pass more test cases (and yes we have more TCs passing in #884 smiley). But please be reassured that I'll purge these annotations after passing the virtual method test cases, and also please be assured that for any additional annotation to be added I'll try to confine them in a function so that it would be easy to indentify the redundant in the future.

Alright, given that this is not a critical part of esbmc, I agree to accept it for now.

But please, remember that: 
(1) esbmc is not well optimized for memory usage, so every extra annotation you add that is not necessary, more memory we are using unnecessarily.
(2) The more annotations you add, the more trouble you'll have when rewriting the frontend to use irep2 in the future. For instance, there is no access annotation in irep2. They are not even migrated in the migration code. I strongly believe that these are not necessary, and you might waste time migrating them to irep2 before getting a PR rejected.

My point is: be wary of your time :)

**Comment by mikhailramalho** [comment]
Also, it's weird that the code style bot didn't complain about the unordered headers, is this option disabled @rafaelsamenezes? Can we enable it?

**Comment by rafaelsamenezes** [comment]
> Also, it's weird that the code style bot didn't complain about the unordered headers

What do you mean by sorted? As in category or alphabetical?

For categories there are some regex that we can add: https://clang.llvm.org/docs/ClangFormatStyleOptions.html#includecategories

Alphabetical is a matter of adding [SortIncludes](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#sortincludes) into .clang-format file.

**Review by rafaelsamenezes — APPROVED** [review]
